allsky_mqttsubscribe/allsky_mqttsubscribe.py

```diff
@@ -0,0 +1,201 @@
+    Returns:
+        str: The result of the MQTT subscription process. It could be the received message,
+             "Invalid JSON" if the message is not a valid JSON, "No message received" if no
+             message was received, or an error message if the connection to the MQTT server failed.
+    """
+    result = ""
+    extra_data = {}
+    extradatafilename = params['extradatafilename']
+    period = int(params['period'])
+    mqtt_server = params['mqttserver']
+    mqtt_port = int(params['mqttport'])
+    mqtt_topic = params['mqtttopic']
+
+    # Check if the module should run
+    should_run, diff = s.shouldRun(metaData['module'], period)
+
+    if should_run:
+        try:
+            # MQTT client callbacks for connection
+            def on_connect(client, userdata, flags, rc):
+                if rc == 0:
+                    s.log(1, f'INFO: Connected to MQTT server {mqtt_server}:{mqtt_port}')
+                    client.subscribe(mqtt_topic)
+                else:
+                    s.log(1, f'ERROR: Connection to MQTT server failed with code {rc}')
+
+            # MQTT client callbacks for message
+            def on_message(client, userdata, msg):
+                nonlocal result
+                nonlocal extra_data
+                try:
+                    payload = msg.payload.decode('utf-8')
+                    s.log(1, f'INFO: Received message: {payload}')
+                    json_data = json.loads(payload)
+                    result = json_data
+                    extra_data = json_data
+                except json.JSONDecodeError as e:
+                    s.log(1, f'ERROR: Failed to decode JSON message: {e}')
+                    result = "Invalid JSON"
+
+            # Create MQTT client and connect to the server
+            client = mqtt.Client()
+            client.on_connect = on_connect
+            client.on_message = on_message
+
+            # Set MQTT username and password if provided
+            if params['mqttusername'] and params['mqttpassword']:
+                client.username_pw_set(params['mqttusername'], params['mqttpassword'])
+
+            client.connect(mqtt_server, mqtt_port, 60)
+
+            # Run the client for a second to connect and receive messages
+            client.loop_start()
+            time.sleep(1)
+            client.loop_stop()
+            client.disconnect()
+
+            # Check if a message was received
+            if not extra_data:
+                result = "No message received"
+
+            # Save the extra data and set the last run
+            s.log(1, f'INFO: Final result: {result}')
+            s.saveExtraData(extradatafilename, extra_data)
+            s.setLastRun(metaData['module'])
+            return result
+
+        # Handle exceptions
+        except Exception as e:
+            s.log(1, f'ERROR: Failed to connect to MQTT server: {e}')
+            return "Failed to connect to MQTT server"
```

Comment:
This should set result and return it rather than return in the middle of the function

Reply:
Fixed this one too.